lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -283,6 +283,37 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
 
   // Propagate the reachable RefSendOp across modules.
   LogicalResult handleInstanceOp(InstanceOp inst) {
+    auto extRefMod =
+        dyn_cast<FExtModuleOp>(instanceGraph->getReferencedModule(inst));
+    if (extRefMod) {
+      // Extern modules can generate RefType ports, they have an attached
+      // attribute which specifies the internal path into the extern module.
+      // This string attribute will be used to generate the final xmr.
+      auto internalPaths = extRefMod.getInternalPaths();
+      if (internalPaths.empty())
+        return success();
+      size_t pathsIndex = 0;
+      auto numPorts = inst.getNumResults();
+      for (const auto &res : llvm::enumerate(inst.getResults())) {
+        if (!inst.getResult(res.index()).getType().isa<RefType>())
+          continue;
+
+        auto inRef = getInnerRefTo(inst);
+        auto ind = addReachingSendsEntry(res.value(), inRef);
+
+        xmrPathSuffix[ind] =
+            "." + internalPaths[pathsIndex].cast<StringAttr>().str();
+        ++pathsIndex;
+        // The instance result and module port must be marked for removal.
+        if (refPortsToRemoveMap[inst].size() < numPorts)
+          refPortsToRemoveMap[inst].resize(numPorts);
+        refPortsToRemoveMap[inst].set(res.index());
+        if (refPortsToRemoveMap[extRefMod].size() < numPorts)
+          refPortsToRemoveMap[extRefMod].resize(numPorts);
+        refPortsToRemoveMap[extRefMod].set(res.index());
```

Comment:
Moving this to a shared lambda or separating this into two commented regions would improve readability.